### PR TITLE
More models

### PR DESCRIPTION
**Pull request by Asbiorn**
Added last models, set their relationships, and i hope i got i right. 
search for TODO tag, to see where I may be in doubt. See the figure DBmodel.png for the changes i've made to the DB setup.

### CODE REVIEW

File: src/main/java/noroff/mefit/models/Address.java [hunk 1]
```diff
@@ -8,6 +8,20 @@ public class Address {
...
 
+    @Column(length = 150, nullable = true)
+    private String addressLine1 ;
+    @Column(length = 150, nullable = true)
+    private String addressLine2;
+    @Column(length = 100, nullable = true)
+    private String addressLine3;
+
+    @Column(length = 100, nullable = true)
+    private String postalCode;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Address.java:19]: postalCode might need to be of type int

File: src/main/java/noroff/mefit/models/Exercise.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import jakarta.persistence.*;
+
+@Entity
+public class Exercise {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int id;
+    @Column(length = 100, nullable = false)
+    private String name;
+    @Column(length = 100, nullable = false)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]:
```suggestion
    @Column(name = "description", length = 100, columnDefinition="TEXT")
```

File: src/main/java/noroff/mefit/models/Exercise.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    private String description;
+
+    @Column(length = 100, nullable = false)
+    private String targetMuscleGroup; //TODO Should this be set of ENUMs?
+
+    @Column(length = 100, nullable = false)
+    private String image; //TODO decide whether to use blob or bytea to store image.
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]: it's generally a bad idea to store images in a relational database. A database is a data storage system for primitive data (numbers, text, booleans etc). A file, like an image, is not a primitive, so it should be stored on dedicated file storages like a file-server, or the hard-drive.

Instead a URL path to the image stored on a harddrive should be provided, so data type should be a string.

Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]:
```suggestion
    private String imageURL; 
```

File: src/main/java/noroff/mefit/models/Exercise.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Column(length = 255, nullable = true)
+    private String vidLink;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]:
```suggestion
    private String videoURL;
```

File: src/main/java/noroff/mefit/models/Exercise.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    @OneToMany(mappedBy = "exercise")
+    private java.util.Set<Set> sets ;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]:
```suggestion
    private Set<ExerciseSet> exerciseSets ;
```

File: src/main/java/noroff/mefit/models/Goal.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Set;
+
+@Entity
+public class Goal {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int id;
+
+    @Column(length = 5, nullable = false)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Goal.java:16]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/noroff/mefit/models/Goal.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    private boolean acheived ;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Goal.java:17]:
```suggestion
    private boolean achieved ;
```

File: src/main/java/noroff/mefit/models/Goal.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Column(name="end_date",length = 100, nullable = false)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Goal.java:19]:
```suggestion
    @Column(name="end_date")
```

File: src/main/java/noroff/mefit/models/Goal.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package noroff.mefit.models;
+
+
+import jakarta.persistence.*;
+
+import java.time.LocalDate;
+import java.util.Set;
+
+@Entity
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Goal.java:9]:
```suggestion
@Entity
@Table(name = "goal")
```

File: src/main/java/noroff/mefit/models/Goal.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    @OneToOne(mappedBy = "goal")
+    private Profile profile;
+
+    // TODO should goals be able to have more than one program?
+    @OneToOne
+    @JoinColumn(name = "program_id")
+    private Program program;
+
+    @ManyToMany
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Goal.java:31]: Not sure if this is the desired relationship. Let's discuss this further

File: src/main/java/noroff/mefit/models/Profile.java [hunk 11]
```diff
@@ -9,13 +9,35 @@ public class Profile {
...
     private int id;
 
+    @Column(length = 5, nullable = false)
+    private float weight ;
+
+    @Column(length = 5, nullable = false)
+    private float height ;
+
+    @Column(length = 255, nullable = true)
+    private String disability;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Profile.java:19]:
```suggestion
    private boolean disability;
```

File: src/main/java/noroff/mefit/models/Profile.java [hunk 12]
```diff
@@ -9,13 +9,35 @@ public class Profile {
...
+
+    @Column(name="medical_condition", length = 255, nullable = true)
+    private String medicalConditions;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Profile.java:22]:
```suggestion
    private boolean medicalConditions;
```

File: src/main/java/noroff/mefit/models/Profile.java [hunk 13]
```diff
@@ -9,13 +9,35 @@ public class Profile {
...
 
     @OneToOne
     @JoinColumn(name = "address_id")
     private Address address;
+
+    @OneToOne
+    @JoinColumn(name = "goal_id")
+    private Goal goal;
+
+    @OneToOne
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Profile.java:38]: I think a user can have many programs?

File: src/main/java/noroff/mefit/models/Program.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.Set;
+
+@Entity
+public class Program {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int id;
+    @Column(length = 100, nullable = false)
+    private String name;
+    @Column(length = 50, nullable = false)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Program.java:14]:
```suggestion
    @Column(name = "program_category")
    @Enumerated(EnumType.STRING)
```

File: src/main/java/noroff/mefit/models/Program.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    private String category;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Program.java:15]:
```suggestion
    private Category program_category;
```

File: src/main/java/noroff/mefit/models/Workout.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    @ManyToMany(mappedBy = "workouts")
+    private Set<Goal> goals;
+
+    @ManyToMany
+    @JoinTable(
+            name = "workout_set",
+            joinColumns = @JoinColumn(name = "workout_id"),
+            inverseJoinColumns = @JoinColumn(name = "set_id")
+    )
+    private Set<noroff.mefit.models.Set> sets ;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Workout.java:34]:
```suggestion
    private Set<ExerciseSet> exerciseSets ;
```

File: src/main/java/noroff/mefit/models/Set.java [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+    @Column(length = 5, nullable = false)
+    private int exercise_repetitions;
+
+    @ManyToOne
+    @JoinColumn(name = "exercise_id")
+    private Exercise exercise;
+
+    @ManyToMany(mappedBy = "sets")
+    private java.util.Set<Workout> workouts;
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Set.java:19]:
```suggestion
    private Set<Workout> workouts;
```

File: src/main/java/noroff/mefit/models/UserAcc.java [hunk 18]
```diff
@@ -3,21 +3,24 @@
...
     @Column(length = 50, nullable = false)
     private String first_name;
-
     @Column(length = 50, nullable = false)
     private String last_name;
-    @Column(length = 10, nullable = false)
+    @Column(length = 10, nullable = true)
     private boolean isContributor;
-    @Column(length = 10, nullable = false)
+    @Column(length = 10, nullable = true)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/UserAcc.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/noroff/mefit/models/UserAcc.java [hunk 19]
```diff
@@ -3,21 +3,24 @@
...
+    private String password;
+
+
     @Column(length = 50, nullable = false)
     private String first_name;
-
     @Column(length = 50, nullable = false)
     private String last_name;
-    @Column(length = 10, nullable = false)
+    @Column(length = 10, nullable = true)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/UserAcc.java:18]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/noroff/mefit/models/Exercise.java [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+public class Exercise {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int id;
+    @Column(length = 100, nullable = false)
+    private String name;
+    @Column(length = 100, nullable = false)
+    private String description;
+
+    @Column(length = 100, nullable = false)
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]:
```suggestion
    @Column(name = "target_muscle_group")
    @Enumerated(EnumType.STRING)
```

File: src/main/java/noroff/mefit/models/Exercise.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    private String targetMuscleGroup; //TODO Should this be set of ENUMs?
```
Comment (kristian-andreasen) [inline src/main/java/noroff/mefit/models/Exercise.java]:
```suggestion
    private Muscles muscle_group;
```

**Review by kristian-andreasen — COMMENTED** [review]
Good job overall. A lot of boolean types has a "length" attribute which is not necessary. We should discuss relationships together I think.

File: src/main/java/noroff/mefit/models/Address.java (hunk 1 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Address.java:19]:
```suggestion
    private int postalCode;
```

File: src/main/java/noroff/mefit/models/Exercise.java (hunk 5 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Exercise.java]: we need to rename Set then as well..

File: src/main/java/noroff/mefit/models/Profile.java (hunk 11 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Profile.java:19]: We don't want a description of these then?

File: src/main/java/noroff/mefit/models/Profile.java (hunk 13 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Profile.java:38]: sure

File: src/main/java/noroff/mefit/models/Program.java (hunk 14 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Program.java:14]: re-review

File: src/main/java/noroff/mefit/models/Program.java (hunk 15 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Program.java:15]: no enums?

File: src/main/java/noroff/mefit/models/Set.java (hunk 17 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/Set.java:19]: wont work like that. until we rename "Set". there's ambiguity when we have a class called "Set" which is also java interface

File: src/main/java/noroff/mefit/models/UserAcc.java (hunk 19 above)
Comment (Asbiorn) [reply src/main/java/noroff/mefit/models/UserAcc.java:18]: I would say that it's not good to have a user with only a first name.

**Review by kristian-andreasen — APPROVED** [review]
nice job

**Pull request merged**